### PR TITLE
[MXNET-472]  ccache for docker builds

### PR DESCRIPTION
**Pull request by lebeg**
## Description ##

This is a work on top of [Perdo's PR](https://github.com/apache/incubator-mxnet/pull/11036), but has a few improvements.

### Essentials ###
- [x] The PR title starts with [MXNET-472](https://issues.apache.org/jira/projects/MXNET/issues/MXNET-472)
- [x] Changes are complete

### Changes ###
- [x] CCache latest version and build for all different docker builds
- [x] CPU and GPU builds are supported

## Comments ##

The ccache dir is meant to be shared via EFS on instances with the same label on CI builds.

### CODE REVIEW

File: ci/docker/Dockerfile.build.android_arm64 [hunk 1]
```diff
@@ -18,9 +18,16 @@
...
 
+FROM ccache/build.ubuntu as ccachebuilder
+
+COPY install/ubuntu_install_ccache.sh /work/
+RUN /work/ubuntu_install_ccache.sh
+
 FROM dockcross/base:latest
 MAINTAINER Pedro Larroy "pllarroy@amazon.com"
 
+COPY --from=ccachebuilder /usr/local/bin/ccache /usr/local/bin/ccache
```
Comment (larroy) [inline ci/docker/Dockerfile.build.android_arm64]: what does this do? can we add a comment?

File: ci/docker/install/centos7_install_ccache.sh [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+yum -y install wget
+yum -y install make
+yum -y install google-perftools
+yum -y install asciidoc
+yum -y install gcc-c++-4.8.*
+
+mkdir -p /work/deps
+cd /work/deps
+
+git clone --recursive -b v3.4.2 https://github.com/ccache/ccache.git
```
Comment (larroy) [inline ci/docker/install/centos7_install_ccache.sh]: nice!

File: ci/docker/install/centos7_install_ccache.sh [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+pushd .
+
+yum -y install epel-release
+yum -y install git
+yum -y install autoconf
+yum -y install wget
+yum -y install make
+yum -y install google-perftools
+yum -y install asciidoc
+yum -y install gcc-c++-4.8.*
```
Comment (larroy) [inline ci/docker/install/centos7_install_ccache.sh]: isn't this part of centos7_core.sh . ? we should not duplicate it I think

File: ci/docker/install/centos7_install_ccache.sh [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+./configure
+make -j$(nproc)
+make install
+
+cd /work/deps
+rm -rf /work/deps/ccache
+
+popd
+
+export CCACHE_MAXSIZE=${CCACHE_MAXSIZE:=10G}
```
Comment (larroy) [inline ci/docker/install/centos7_install_ccache.sh]: who will pick up these variables? Shouldn't it be in the dockerfile or in the python script?

File: ci/docker/install/ubuntu_install_ccache.sh [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+# Script to build ccache for ubuntu based images
+
+set -ex
+
+pushd .
+
+apt update
+apt install -y --no-install-recommends \
+    git \
```
Comment (larroy) [inline ci/docker/install/ubuntu_install_ccache.sh]: same comment above about "core"

File: ci/docker/runtime_functions.sh [hunk 6]
```diff
@@ -31,14 +31,49 @@ clean_repo() {
     git submodule update --init --recursive
 }
 
+# wrap compiler calls with ccache
+build_ccache_wrappers() {
```
Comment (larroy) [inline ci/docker/runtime_functions.sh]: do we need wrappers if we compile the latest ccache ourselves? it should have the fix for nvcc right?

Comment (lebeg) [reply ci/docker/runtime_functions.sh]: Yes, a fix has been added as can be seen in [release notes for 3.4](https://github.com/ccache/ccache.git) but this is unfortunately not enough, I couldn't get in working.

File: ci/docker/install/centos7_install_ccache.sh (hunk 4 above)
Comment (lebeg) [reply ci/docker/install/centos7_install_ccache.sh]: I was thinking about it but decided to keep it here - it has the advantage to configure what was installed in 1 place. And actually this is setting only a default value if variable was not set before by the script.

File: ci/docker/install/centos7_install_ccache.sh (hunk 3 above)
Comment (lebeg) [reply ci/docker/install/centos7_install_ccache.sh]: This script is called sometimes separately from install core.

File: ci/docker/Dockerfile.build.android_arm64 (hunk 1 above)
Comment (lebeg) [reply ci/docker/Dockerfile.build.android_arm64]: Will do.

**Comment by larroy** [comment]
I would prefer to try this functionality without EFS first, to asses EFS impact.

**Comment by marcoabreu** [comment]
Agree

File: ci/docker/install/centos7_install_ccache.sh (hunk 3 above)
Comment (marcoabreu) [reply ci/docker/install/centos7_install_ccache.sh]: installing a compiler should not be in this file

File: ci/docker/install/centos7_install_ccache.sh (hunk 4 above)
Comment (marcoabreu) [reply ci/docker/install/centos7_install_ccache.sh]: I think we should set it in build.py, considering we're bootstrapping ccache there anyways and it allows a centralized configuration. With the current way, you'd have to change it in multiple scripts (ubuntu, centos, windows, arm, etc)

File: ci/build.py [hunk 7]
```diff
@@ -116,9 +119,19 @@ def buildir() -> str:
     return os.path.join(get_mxnet_root(), "build")
 
 
+def default_ccache_dir() -> str:
+    if 'CCACHE_DIR' in os.environ:
+        ccache_dir = os.path.realpath(os.environ['CCACHE_DIR'])
+        os.makedirs(ccache_dir, exist_ok=True)
+        return ccache_dirpython
+    # Share ccache across containers (should we have a separate dir per platform?)
```
Comment (marcoabreu) [inline ci/build.py]: No, ccache should be able to figure this out on its own, right?

File: ci/build.py [hunk 8]
```diff
@@ -127,12 +140,16 @@ def container_run(platform: str,
...
-        '-v', "{}:/work/mxnet".format(mx_root), # mount mxnet root
-        '-v', "{}:/work/build".format(local_build_folder), # mount mxnet/build for storing build artifacts
-        '-u', '{}:{}'.format(os.getuid(), os.getgid()),
-        tag]
+               '--shm-size={}'.format(shared_memory_size),
+               '-v', "{}:/work/mxnet".format(mx_root),  # mount mxnet root
+               '-v', "{}:/work/build".format(local_build_folder),  # mount mxnet/build for storing build artifacts
+               '-v', "{}:/work/ccache".format(local_ccache_dir),
+               '-u', '{}:{}'.format(os.getuid(), os.getgid()),
+               '-e', "CCACHE_DIR=/work/ccache",  # this path is inside the container as /work/ccache is mounted
```
Comment (marcoabreu) [inline ci/build.py]: Add cache-size limit here

File: ci/docker/Dockerfile.build.jetson [hunk 9]
```diff
@@ -22,13 +22,21 @@
...
 # FROM dockcross/linux-arm64
 FROM mxnetci/dockcross-linux-arm64:05082018
 
+# extract ccache binary into latest context
+COPY --from=ccachebuilder /usr/local/bin/ccache /usr/local/bin/ccache
+
 ENV ARCH aarch64
-ENV FC /usr/bin/${CROSS_TRIPLE}-gfortran
 ENV HOSTCC gcc
+ENV FC /usr/bin/${CROSS_TRIPLE}-gfortran
```
Comment (marcoabreu) [inline ci/docker/Dockerfile.build.jetson]: How come FC is required here but not for the ARMv6 dockerfile?

File: ci/docker/Dockerfile.build.jetson [hunk 10]
```diff
@@ -39,6 +47,9 @@ RUN git clone --recursive -b v0.2.20 https://github.com/xianyi/OpenBLAS.git && \
     make -j$(nproc) && \
     PREFIX=${CROSS_ROOT} make install
 
+ENV OpenBLAS_HOME=${CROSS_ROOT}
+ENV OpenBLAS_DIR=${CROSS_ROOT}
```
Comment (marcoabreu) [inline ci/docker/Dockerfile.build.jetson]: Unrelated change?

File: ci/docker/install/centos7_install_ccache.sh [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+# specific language governing permissions and limitations
+# under the License.
+
+# Script to build ccache for centos7 based images
+
+set -ex
+
+pushd .
+
+yum -y install epel-release
```
Comment (marcoabreu) [inline ci/docker/install/centos7_install_ccache.sh]: Please make sure you're only installing necessary dependencies. It seems like quite a lot are duplicate or just simply don't fit here. Some examples would be git, wget, make, asciidoc, gcc. For the others, I'm not sure what they exactly do.

File: ci/docker/install/centos7_install_ccache.sh [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+make -j$(nproc)
+make install
+
+cd /work/deps
+rm -rf /work/deps/ccache
+
+popd
+
+export CCACHE_MAXSIZE=${CCACHE_MAXSIZE:=10G}
+export CCACHE_DIR=${CCACHE_DIR:=/work/ccache}
```
Comment (marcoabreu) [inline ci/docker/install/centos7_install_ccache.sh]: I don't think the environment variable is going to persist this way.

File: ci/docker/install/ubuntu_install_ccache.sh (hunk 5 above)
Comment (marcoabreu) [reply ci/docker/install/ubuntu_install_ccache.sh]: +1

File: ci/docker/runtime_functions.sh [hunk 13]
```diff
@@ -31,14 +31,49 @@ clean_repo() {
     git submodule update --init --recursive
 }
 
+# wrap compiler calls with ccache
```
Comment (marcoabreu) [inline ci/docker/runtime_functions.sh]: Please add more documentation. To me, it's not clear what exactly is happening here and why we have to create these files in our source directory

File: ci/docker/runtime_functions.sh [hunk 14]
```diff
@@ -31,14 +31,49 @@ clean_repo() {
...
+    fi
+
+    if [ -z ${CXX+x} ]; then
+       echo "No \$CXX set, defaulting to g++";
+       export CXX=g++
+    fi
+
+    # this function is nessesary for cuda enabled make based builds, since nvcc needs just an executable for -ccbin
+
+    echo -e "#!/bin/sh\n/usr/local/bin/ccache ${CC} \"\$@\"\n" >> cc
```
Comment (marcoabreu) [inline ci/docker/runtime_functions.sh]: why use touch and >> instead of a simple >?

File: ci/docker/runtime_functions.sh [hunk 15]
```diff
@@ -31,14 +31,49 @@ clean_repo() {
...
 # function here with the same suffix:
 
 build_jetson() {
     set -ex
     pushd .
-    mv make/crosscompile.jetson.mk make/config.mk
+
+    build_ccache_wrappers
+
+    cp -f make/crosscompile.jetson.mk ./config.mk
```
Comment (marcoabreu) [inline ci/docker/runtime_functions.sh]: This is a different path now, right? Was the first one wrong?

File: ci/docker/runtime_functions.sh (hunk 13 above)
Comment (marcoabreu) [reply ci/docker/runtime_functions.sh]: Also, is this only required before making make-calls? It seems like you're not using it for cmake. So maybe rename the function into something that makes it clear that this is a make-specific thing

File: ci/docker/runtime_functions.sh [hunk 16]
```diff
@@ -178,6 +225,8 @@ build_centos7_cpu() {
 build_centos7_mkldnn() {
     set -ex
     cd /work/mxnet
+    export CC="ccache gcc"
+    export CXX="ccache g++"
```
Comment (marcoabreu) [inline ci/docker/runtime_functions.sh]: How come you don't use the wrapper here?

File: ci/docker/runtime_functions.sh [hunk 17]
```diff
@@ -190,6 +239,8 @@ build_centos7_mkldnn() {
 build_centos7_gpu() {
     set -ex
     cd /work/mxnet
+    # unfortunately this build has problems in 3rdparty dependencies with ccache and make
+    # build_ccache_wrappers
```
Comment (marcoabreu) [inline ci/docker/runtime_functions.sh]: Could you elaborate?

File: make/crosscompile.jetson.mk [hunk 18]
```diff
@@ -96,7 +96,7 @@ USE_LIBJPEG_TURBO = 0
 USE_LIBJPEG_TURBO_PATH = NONE
 
 # use openmp for parallelization
-USE_OPENMP = 1
+USE_OPENMP = 0
```
Comment (marcoabreu) [inline make/crosscompile.jetson.mk]: Please change back to 1

File: ci/build.py (hunk 8 above)
Comment (larroy) [reply ci/build.py]: default is 5G I think that's ok for now. We can tune it later.

File: ci/docker/Dockerfile.build.jetson (hunk 10 above)
Comment (larroy) [reply ci/docker/Dockerfile.build.jetson]: I also don't understand this one, can we add a comment on what do those variables do?

File: ci/docker/install/centos7_install_ccache.sh (hunk 4 above)
Comment (larroy) [reply ci/docker/install/centos7_install_ccache.sh]: I don't think this variable gets exported outside of this script. So it won't be used. Am I wrong?

File: ci/docker/install/centos7_install_ccache.sh (hunk 12 above)
Comment (larroy) [reply ci/docker/install/centos7_install_ccache.sh]: Agree with @marcoabreu on this one, at least that's my experience.

File: ci/docker/runtime_functions.sh (hunk 14 above)
Comment (larroy) [reply ci/docker/runtime_functions.sh]: Agree, I think we should use >   doesn't make sense to append in case the file already exist does it?

File: ci/docker/install/centos7_install_ccache.sh (hunk 3 above)
Comment (lebeg) [reply ci/docker/install/centos7_install_ccache.sh]: In this file only requirements to build ccache are installed. It works after calling install core or on it's own.

File: ci/docker/install/centos7_install_ccache.sh (hunk 4 above)
Comment (lebeg) [reply ci/docker/install/centos7_install_ccache.sh]: In the current way you change it only in this script. And the way it's done does not prevent you to have it in build.py as well.

File: ci/docker/install/centos7_install_ccache.sh (hunk 3 above)
Comment (marcoabreu) [reply ci/docker/install/centos7_install_ccache.sh]: When is it called separately from install core? These scripts are not supposed to be standalone but only serve what they are intended to do.

Comment (lebeg) [reply ci/docker/install/centos7_install_ccache.sh]: It is called separately for all cross compilation platforms.

File: ci/build.py (hunk 7 above)
Comment (lebeg) [reply ci/build.py]: Maybe

File: ci/build.py (hunk 8 above)
Comment (lebeg) [reply ci/build.py]: Actually the default is in the install script, see my comments above

File: ci/docker/Dockerfile.build.jetson (hunk 9 above)
Comment (lebeg) [reply ci/docker/Dockerfile.build.jetson]: Because I've fixed it for [dockcross/master](https://github.com/dockcross/dockcross/pull/228), but we can not use dockcross master everywhere - for some reason CUDA builds are broken. So we need to stay with your workaround fork, which does not contain this fix.

File: ci/docker/Dockerfile.build.jetson (hunk 10 above)
Comment (lebeg) [reply ci/docker/Dockerfile.build.jetson]: This is important for [FindOpenBLAS.cmake](https://github.com/apache/incubator-mxnet/blob/master/cmake/Modules/FindOpenBLAS.cmake#L49-L52) it is not used here because it's still a makefile build, but will be used when the build will be ported.

File: ci/docker/install/centos7_install_ccache.sh (hunk 11 above)
Comment (lebeg) [reply ci/docker/install/centos7_install_ccache.sh]: Everything listed here is needed for an isolated ccache build from source.

File: ci/docker/install/centos7_install_ccache.sh (hunk 12 above)
Comment (lebeg) [reply ci/docker/install/centos7_install_ccache.sh]: Hm, you might be right here.

File: ci/docker/runtime_functions.sh (hunk 13 above)
Comment (lebeg) [reply ci/docker/runtime_functions.sh]: There is a comment about this 2 lines below. For cmake it is not nessesary

File: ci/docker/runtime_functions.sh (hunk 14 above)
Comment (lebeg) [reply ci/docker/runtime_functions.sh]: Yes, this was the first thing I have tried, but I had problems and could not get it to work. It seemed that EOF was not what it should be.

File: ci/docker/runtime_functions.sh (hunk 15 above)
Comment (lebeg) [reply ci/docker/runtime_functions.sh]: This is the preferred way of doing it without causing git tracked changes on every build

File: ci/docker/runtime_functions.sh (hunk 16 above)
Comment (lebeg) [reply ci/docker/runtime_functions.sh]: At first I have tried it and unfortunately the build breaks due to make build of dependencies (mshadow). This was a necessary revert to Pedros solution.

File: ci/docker/install/centos7_install_ccache.sh (hunk 3 above)
Comment (marcoabreu) [reply ci/docker/install/centos7_install_ccache.sh]: Well but the cross compilation platforms also bring their own compilers, right? I think we should really doublecheck what's actually related to ccache and what's a better fit for another place

File: ci/docker/install/centos7_install_ccache.sh (hunk 4 above)
Comment (marcoabreu) [reply ci/docker/install/centos7_install_ccache.sh]: The problem is that this environment variable will not persist, means this call is a NO-OP.

File: ci/docker/Dockerfile.build.jetson (hunk 9 above)
Comment (marcoabreu) [reply ci/docker/Dockerfile.build.jetson]: Ah I see, that's a great contribution! Thank you

File: ci/docker/install/centos7_install_ccache.sh (hunk 11 above)
Comment (marcoabreu) [reply ci/docker/install/centos7_install_ccache.sh]: See discussion below.

File: ci/docker/install/centos7_install_ccache.sh (hunk 3 above)
Comment (marcoabreu) [reply ci/docker/install/centos7_install_ccache.sh]: I really don't think that installing utilities like asciidoc, wget, git or all that stuff is actually related to ccache.

File: ci/docker/runtime_functions.sh (hunk 15 above)
Comment (marcoabreu) [reply ci/docker/runtime_functions.sh]: I was just confused about make/config.mk -> ./config.mk

File: ci/docker/install/centos7_install_ccache.sh (hunk 3 above)
Comment (lebeg) [reply ci/docker/install/centos7_install_ccache.sh]: Here are [ccache installation instructions](https://github.com/ccache/ccache/blob/master/doc/INSTALL.md).

Comment (lebeg) [reply ci/docker/install/centos7_install_ccache.sh]: The compilers for cross compilation will be in a different head (different FROM)

File: ci/docker/install/centos7_install_ccache.sh (hunk 4 above)
Comment (lebeg) [reply ci/docker/install/centos7_install_ccache.sh]: Yes, I agree that might be a problem

**Comment by marcoabreu** [comment]
Replaced with https://github.com/apache/incubator-mxnet/pull/11151
